src/tools/w3dviewer/emitterlineproppage.cpp

```diff
@@ -0,0 +1,153 @@
+
+    if (ud != nullptr && ud->hdr.code == UDN_DELTAPOS) {
+        UpdateEditCtrl(ud->hdr.hwndFrom, ud->iDelta);
+    }
+
+    return CWnd::OnNotify(wParam, lParam, pResult);
+}
+
+void EmitterLinePropPageClass::DoDataExchange(CDataExchange *pDX)
+{
+    DDX_Control(pDX, IDC_MAPPING, m_mapping);
+    DDX_Control(pDX, IDC_MERGEABORTSPIN, m_mergeAbortSpin);
+    DDX_Control(pDX, IDC_VPERSECSPIN, m_vPerSecSpin);
+    DDX_Control(pDX, IDC_UVTILESPIN, m_uvTileSpin);
+    DDX_Control(pDX, IDC_UPERSECSPIN, m_uPerSecSpin);
+    DDX_Control(pDX, IDC_AMPLITUDESPIN, m_amplitudeSpin);
+    DDX_Control(pDX, IDC_SUBDIVISIONSPIN, m_subdivisionSpin);
+}
+
+BOOL EmitterLinePropPageClass::OnInitDialog()
+{
+    CDialog::OnInitDialog();
+    m_subdivisionSpin.SetRange(0, 8);
+    m_subdivisionSpin.SetPos(m_subdivisionLevel);
+    InitializeSpinButton(&m_amplitudeSpin, m_noiseAmplitude, -10000.0f, 10000.0f);
+    InitializeSpinButton(&m_mergeAbortSpin, m_mergeAbortFactor, -10000.0f, 10000.0f);
+    InitializeSpinButton(&m_uvTileSpin, m_textureTileFactor, 0.0f, 8.0f);
+    InitializeSpinButton(&m_uPerSecSpin, m_uPerSec, 0.0f, 32.0f);
+    InitializeSpinButton(&m_vPerSecSpin, m_vPerSec, 0.0f, 32.0f);
+    m_mapping.SetCurSel(m_textureMappingMode);
+    SendDlgItemMessage(IDC_MERGE, BM_SETCHECK, m_mergeIntersections, 0);
+    SendDlgItemMessage(IDC_ENDCAPS, BM_SETCHECK, m_endCaps, 0);
+    SendDlgItemMessage(IDC_SORTING, BM_SETCHECK, m_disableSorting, 0);
+    DisableWindows(m_hWnd, m_instanceList->Get_Render_Mode() == W3D_EMITTER_RENDER_MODE_LINE);
+    return TRUE;
+}
+
+BOOL EmitterLinePropPageClass::OnApply()
+{
+    m_subdivisionLevel = GetDlgItemInt(IDC_SUBDIVISION);
+    m_noiseAmplitude = GetDlgItemFloat(m_hWnd, IDC_AMPLITUDE);
+    m_mergeAbortFactor = GetDlgItemFloat(m_hWnd, IDC_MERGEABORT);
+    m_textureTileFactor = GetDlgItemFloat(m_hWnd, IDC_UVTILE);
+    m_uPerSec = GetDlgItemFloat(m_hWnd, IDC_UPERSEC);
+    m_vPerSec = GetDlgItemFloat(m_hWnd, IDC_VPERSEC);
+    m_textureMappingMode = SendDlgItemMessage(IDC_MAPPING, CB_GETCURSEL, 0, 0);
+    m_mergeIntersections = SendDlgItemMessage(IDC_MERGE, BM_GETCHECK, 0, 0) != 0;
+    m_endCaps = SendDlgItemMessage(IDC_ENDCAPS, BM_GETCHECK, 0, 0) != 0;
+    m_disableSorting = SendDlgItemMessage(IDC_SORTING, BM_GETCHECK, 0, 0) != 0;
+    m_instanceList->Set_Subdivision_Level(m_subdivisionLevel);
+    m_instanceList->Set_Noise_Amplitude(m_noiseAmplitude);
+    m_instanceList->Set_Merge_Abort_Factor(m_mergeAbortFactor);
+    m_instanceList->Set_Texture_Tile_Factor(m_textureTileFactor);
+    m_instanceList->Set_UV_Offset_Rate(Vector2(m_uPerSec, m_vPerSec));
+    m_instanceList->Set_Line_Texture_Mapping_Mode(m_textureMappingMode);
+    m_instanceList->Set_Merge_Intersections(m_mergeIntersections);
+    m_instanceList->Set_Disable_Sorting(m_disableSorting);
+    m_instanceList->Set_End_Caps(m_endCaps);
+    return CPropertyPage::OnApply();
+}
+
+void EmitterLinePropPageClass::Initialize()

```

Comment:
Initialize is not called from inside class yeah?

Reply:
Yeah.